discoverd/health/register_test.go

```diff
@@ -0,0 +1,295 @@
+				case <-stop:
+					errCh <- false
+				}
+			}()
+		}
+
+		var stop chan bool
+		currentMeta := make(map[string]string)
+		for _, step := range steps {
+			stop = make(chan bool)
+			var wait int
+
+			if step.event {
+				monitorChan <- step.up
+			}
+			if step.register {
+				call := <-registrarChan
+				c.Assert(call.args["inst"].(*discoverd.Instance).Meta, DeepEquals, currentMeta)
+				call.returnVal <- step.success
+				<-registrarChan
+			} else {
+				wait++
+				errCheck(registrarChan, stop)
+			}
+			if step.unregister {
+				// before unregistering, Addr should not be nil
+				c.Assert(hb.Addr(), Not(Equals), "")
+				select {
+				case <-unregisterChan:
+				case <-time.After(3 * time.Second):
+					c.Error("Timed out waiting for unregistration")
+				}
+				// Addr should be nil now
+				c.Assert(hb.Addr(), Equals, "")
+			} else {
+				wait++
+				errCheck(unregisterChan, stop)
+			}
+			if step.setMeta {
+				go func() {
+					call := <-metaChan
+					call.returnVal <- step.success
+				}()
+				// SetMeta needs to succeed every time, regardless of the situation
+				currentMeta["TEST"] = random.UUID()
+				c.Assert(hb.SetMeta(currentMeta), IsNil)
+			} else {
+				wait++
+				errCheck(metaChan, stop)
+			}
+
+			// check event forwarding
+			if step.event {
+				select {
+				case val := <-reg.Events:
+					if step.up {
+						c.Assert(val.Status, Equals, MonitorStatusUp)
+					} else {
+						c.Assert(val.Status, Equals, MonitorStatusDown)
+					}
+				case <-time.After(3 * time.Second):
+					c.Error("Timed out waiting for Registration.Events")
+				}
+			}
+
+			// check whether functions were called on this step that should not
+			// have ran
+			close(stop)
+			for i := 0; i < wait; i++ {
+				select {
+				case err := <-errCh:
+					if err {
+						c.Error("Received data on a channel that should not be recieving on this step")
+					}
+				}
+			}
+		}
+		close(unregisterChan)
+		close(monitorChan)
+		close(registrarChan)
+		close(metaChan)
+	}
+
+	for _, t := range []struct {
+		name  string
+		steps []step
+	}{
+		{
+			name: "register success up/down/up",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{event: true, unregister: true},
+				{event: true, up: true, register: true, success: true},
+			},
+		},
+		{
+			name: "register fails then succeeds",
+			steps: []step{
+				{event: true, up: true, register: true, success: false},
+				{register: true, success: true},
+			},
+		},
+		{
+			name: "register is called only once if we get two up events",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{event: true, up: true, register: false},
+			},
+		},
+		{
+			name: "setmeta while registered",
+			steps: []step{
+				{event: true, up: true, register: true, success: true},
+				{setMeta: true, success: true},
+			},
+		},
+		{
+			name: "setmeta while offline",
+			steps: []step{
+				{setMeta: true, success: false},
+			},
```

Comment:
It think this also needs an extra step or two. `setMeta` while offline should just save locally, it should not call the underlying `SetMeta` func